### PR TITLE
Optimize logging

### PR DESCRIPTION
**Pull request by domsj**
This pull request fixes the crash_logging and implements the optimization in which strings for the crashlogger are only computed when needed.
The lazyness is implemented by the type Crash_logger.msg which is either an immediate string or a (unit -> string).
Some syntactic sugar for this has been added in the form macros.

Logger.{debug/info/...}_f section "format%s string" "bla"
->
Logger.log_ Logger.{Debug/Info/...} section (fun () -> Printf.sprintf "format%s string" "bla")

macros ending with _ also capture the variable 'section' which is expected to be defined in the environment, e.g.:

Logger.{debug/info/...}_f_ "format%s string" "bla"
->
Logger.log_ Logger.{Debug/Info/...} section (fun () -> Printf.sprintf "format%s string" "bla")

I tried to eliminate all usages of Lwt_log (except in the modules specifically dealing with logging ofcourse). Some usages are still left but these are not statements doing logging themselves; if desired I can centralize this into the logger module too.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -9,6 +9,7 @@ all: build
 clean:
 	ocamlbuild -clean
 
+
```
Comment (NicolasT) [inline Makefile]: Revert?

File: logger_macro.ml [hunk 2]
```diff
@@ -0,0 +1,168 @@
+open Camlp4.PreCast
```
Comment (NicolasT) [inline logger_macro.ml:1]: Might be useful to give an overview of what gets rewritten into what. It's not entirely clear/obvious if you ask me (yet I don't know camlp4, so that might be part of the issue).

File: src/client/client_protocol.ml [hunk 3]
```diff
@@ -62,9 +61,11 @@ let read_command (ic,oc) =
       end
   end
 
+let log_debug m =
+  Logger.debug_ m
```
Comment (NicolasT) [inline src/client/client_protocol.ml:65]: Does this still rework as expected when macro rewriting is in place?

File: Makefile (hunk 1 above)
Comment (domsj) [inline Makefile]: I changed this file in 2 different commits, resulting in the net addition of 1 new line.
Do you prefer another commit removing this extra line or should I do some history rewriting and collapse everything into 1 commit (and undo this 1 line in the process)?

File: logger_macro.ml (hunk 2 above)
Comment (domsj) [inline logger_macro.ml:1]: Ok, I'll add the examples from the explanation in this pull request and expand on them some more.

File: src/client/client_protocol.ml (hunk 3 above)
Comment (domsj) [inline src/client/client_protocol.ml:65]: Yes, this should work as desired.
Macros with a format string can't be expressed like that. It does require the m argument here for it to work.
As mentioned in my comment above, I'll add some more comments to the macro file which should make clearer what will work and what won't.

File: Makefile (hunk 1 above)
Comment (NicolasT) [inline Makefile]: Rewrite some history ;-)

**Comment by NicolasT** [comment]
Next to minor comments, LGTM.
